src/translate.rs

```diff
@@ -1162,38 +1180,77 @@ impl FromGlib<i32> for Option<u64> {
     }
 }
 
-/// Translate from a pointer type without taking ownership, transfer: none.
+/// Translate from a pointer type which is annotated with `transfer none`.
+/// The resulting value is referenced at least once, by the bindings.
+///
+/// This is suitable for floating references, which become strong references.
+/// It is also suitable for acquiring non-gobject values, like `gchar*`.
+///
+/// The implementation of this trait should acquire a reference to the value
+/// in a way appropriate to the type, e.g. by referencing or copying.
+/// Values obtained using this trait must be properly released on `drop()`
+/// by the implementing type.
+///
+/// For more information, refer to module level documentation.
 pub trait FromGlibPtrNone<P: Ptr>: Sized {
     unsafe fn from_glib_none(ptr: P) -> Self;
 }
 
-/// Translate from a pointer type taking ownership, transfer: full.
+/// Translate from a pointer type which is annotated with `transfer full`.
+/// This transfers the ownership of the value to the Rust side.
+///
+/// Because ownership can only be transferred if something is already referenced,
+/// this is unsuitable for floating references.
+///
+/// The implementation of this trait should not acquire any references to the value.
+/// Values obtained using this trait must be properly released on `drop()`
+/// by the implementing type.
+///
+/// For more information, refer to module level documentation.
 pub trait FromGlibPtrFull<P: Ptr>: Sized {
     unsafe fn from_glib_full(ptr: P) -> Self;
 }
 
-/// Translate from a pointer type by borrowing. Don't increase the refcount
+/// Translate from a pointer type by borrowing, without affecting the refcount.
+///
+/// The purpose of this trait is to access values inside callbacks
+/// without changing their reference status.
+/// The obtained borrow must not be accessed outside of the scope of the callback.
+/// Care must be taken not to change the reference count of the borrowed value,
+/// and not to let any called procedure to store a reference.
+/// In particular, safe Rust code should never obtain a mutable borrow.
+///
+/// The implementation of this trait must not acquire any references to the value.
```

Comment:
```suggestion
/// The implementation of this trait must not acquire any references to the value or copy it.
```

Reply:
Doesn't `clone()`, as you mentioned above, acquire a reference?

Reply:
It increases the reference count, yes

Reply:
Then I'm seeing a contradiction:

> It's perfectly fine (and often required) to call clone() on the value

but

> /// The implementation of this trait must not acquire any references to the value

How do those two statements work together?

Reply:
The first statement is for safe Rust code that gets a reference (`&T`) to the value passed. It can safely `clone()` if it wants.

The second is about the implementation of the trait. That should not copy or increase the reference count, but just "borrow" the value.

Reply:
Okay, the first sentence is about what can be done on the resulting value, and the second is about the transformation done by the trait. I'm not sure why the two cases are separate though.

Reply:
One is the implementation of the trait and how it should behave. The other is about what is allowed and not allowed with the resulting value (you must not pass it by value or mutable reference to safe code).

Reply:
From the position of the contract of this trait, I think deep-copying the value is always allowed, because that doesn't affect what can be done with the produced reference-thing. Would you agree?

Reply:
Yes, that's the case for all 3 traits. But in case of full/borrow it should not do that unless there's no other way. It would be unexpected to have the overhead of a copy.